plugins/experimental/icap/icap_plugin.cc

```diff
@@ -0,0 +1,1267 @@
+  output_reader                  = NULL;
+  output_vc                      = NULL;
+  output_vio                     = NULL;
+  pending_action                 = NULL;
+  icap_vc                        = NULL;
+  icap_vio                       = NULL;
+  http_body_chunk_length         = -1;
+  http_body_total_length_written = 0;
+  icap_header                    = std::string();
+  http_header                    = std::string();
+  chunk_length_str               = std::string();
+  icap_reply_content_length      = 0;
+  eos_detected                   = false;
+  err_msg                        = std::string();
+  server_reply_content_length    = 0;
+}
+
+TransformData::~TransformData()
+{
+  if (input_reader) {
+    TSIOBufferReaderFree(input_reader);
+  }
+  if (input_buf) {
+    TSIOBufferDestroy(input_buf);
+  }
+  if (os_resp_reader) {
+    TSIOBufferReaderFree(os_resp_reader);
+  }
+  if (os_resp_buf) {
+    TSIOBufferDestroy(os_resp_buf);
+  }
+  if (icap_resp_reader) {
+    TSIOBufferReaderFree(icap_resp_reader);
+  }
+  if (icap_resp_buf) {
+    TSIOBufferDestroy(icap_resp_buf);
+  }
+  if (output_reader) {
+    TSIOBufferReaderFree(output_reader);
+  }
+  if (output_buf) {
+    TSIOBufferDestroy(output_buf);
+  }
+  if (pending_action) {
+    TSActionCancel(pending_action);
+  }
+  if (icap_vc) {
+    TSVConnAbort(icap_vc, 1);
+  }
+}
+
+/*
+ * get_port
+ * Description: Return the port of a sockaddr
+ */
+uint16_t
+get_port(sockaddr const *s_sockaddr)
+{
+  switch (s_sockaddr->sa_family) {
+  case AF_INET: {
+    const struct sockaddr_in *s_sockaddr_in = reinterpret_cast<const struct sockaddr_in *>(s_sockaddr);
+    return ntohs(s_sockaddr_in->sin_port);
+  } break;
+  case AF_INET6: {
+    const struct sockaddr_in6 *s_sockaddr_in6 = reinterpret_cast<const struct sockaddr_in6 *>(s_sockaddr);
+    return ntohs(s_sockaddr_in6->sin6_port);
+  } break;
+  default:
+    return 0;
+    break;
+  }
+}
+
+/*
+ * setup_icap_status_header (Used only in debug-mode)
+ * Description: This function is called to add a customized header
+ *              indicating ICAP server status for logging.
+ */
+static void
+setup_icap_status_header(TransformData *data, const char *header, const char *value)
+{
+  TSMBuffer bufp;
+  TSMLoc resp_loc, field_loc;
+
+  if (TSHttpTxnTransformRespGet(data->txn, &bufp, &resp_loc) != TS_SUCCESS) {
+    TSError("[%s] Couldn't retrieve transform response header", PLUGIN_NAME);
+    return;
+  }
+
+  if (TSMimeHdrFieldCreate(bufp, resp_loc, &field_loc) != TS_SUCCESS) {
+    TSError("[%s] Unable to create field", PLUGIN_NAME);
+    TSHandleMLocRelease(bufp, TS_NULL_MLOC, resp_loc);
+    return;
+  }
+
+  TSMimeHdrFieldNameSet(bufp, resp_loc, field_loc, header, strlen(header));
+  TSMimeHdrFieldValueStringInsert(bufp, resp_loc, field_loc, 0, value, strlen(value));
+  TSMimeHdrFieldAppend(bufp, resp_loc, field_loc);
+
+  TSHandleMLocRelease(bufp, resp_loc, field_loc);
+  TSHandleMLocRelease(bufp, TS_NULL_MLOC, resp_loc);
+}
```

Comment:
If https://github.com/apache/trafficserver/pull/5812 had been merged the body of this function could be:
```
{
  atscppapi::TxnMsg<TSHttpTxnTransformRespGet, atscppapi::RespMsg> msg;

  if (!msg.init(data->txn)) {
    TSError("[%s] Couldn't retrieve transform response header", PLUGIN_NAME);
    return;
  }

  auto fld = atscppapi::MimeField::create(msg, std::string_view(header));

  if (!fld.valid()) {
    TSError("[%s] Unable to create field", PLUGIN_NAME);
    return;
  }
  fld.valuesSet(std::string_view(value));
}
```